GeosCore/fullchem_mod.F90

```diff
@@ -389,131 +455,1262 @@ SUBROUTINE Do_FullChem( Input_Opt,  State_Chm, State_Diag,                 &
-    ! probably safe to define them here outside the OpenMP loop.
-    ! (bmy, 3/28/16)
-    !
-    ! ICNTRL and RCNTRL now need to be updated within the TimeLoop for
-    ! the Rosenbrock_AutoReduce solver (hplin, 4/13/22)
-    !========================================================================
-    DT   = GET_TS_CHEM()          ! Chemistry timestep [s]
-    ATOL = State_Chm%KPP_AbsTol   ! Absolute tolerance
-    RTOL = State_Chm%KPP_RelTol   ! Relative tolerance
+          ! Archive OH reactivity diagnostic
+          CALL Get_OHreactivity ( C, RCONST, OHreact )
+          IF ( State_Diag%Archive_OHreactivity ) THEN
+             State_Diag%OHreactivity(I,J,L) = OHreact
+          ENDIF
+          IF ( State_Diag%Archive_SatDiagnOHreactivity ) THEN
+             State_Diag%SatDiagnOHreactivity(I,J,L) = OHreact
+          ENDIF
+
+       ENDIF
+    ENDDO
+    ENDDO
+    ENDDO
+    !$OMP END PARALLEL DO
 
     !=======================================================================
-    ! %%%%% SOLVE CHEMISTRY -- This is the main KPP solver loop %%%%%
+    ! Return gracefully if integration failed 2x anywhere
+    ! (as we cannot break out of a parallel DO loop!)
     !=======================================================================
-100 format('No. of function calls:', i6, /,                                 &
-           'No. of jacobian calls:', i6, /,                                 &
-           'No. of steps:         ', i6, /,                                 &
-           'No. of accepted steps:', i6, /,                                 &
-           'No. of rejected steps ', i6, /,                                 &
-           '       (except at very beginning)',          /,                 &
-           'No. of LU decompositions:             ', i6, /,                 &
-           'No. of forward/backward substitutions:', i6, /,                 &
-           'No. of singular matrix decompositions:', i6, /,                 &
-            /,                                                              &
-           'Texit, the time corresponding to the      ',        /,          &
-           '       computed Y upon return:            ', f11.4, /,          &
-           'Hexit, last accepted step before exit:    ', f11.4, /,          &
-           'Hnew, last predicted step (not yet taken):', f11.4 )
-
-    !------------------------------------------------------------------------
-    ! Always consider halogens as "fast" species for auto-reduce
-    !------------------------------------------------------------------------
-#ifdef KPP_INTEGRATOR_AUTOREDUCE
-    IF ( FIRSTCHEM .and. Input_Opt%ITS_A_FULLCHEM_SIM ) THEN
-       IF ( Input_Opt%AutoReduce_Is_KeepActive ) THEN
-          CALL fullchem_AR_KeepHalogensActive( Input_Opt%amIRoot )
-       ENDIF
+    IF ( Failed2x ) THEN
+       ErrMsg = 'KPP failed to converge after 2 iterations!'
+       CALL GC_Error( ErrMsg, RC, ThisLoc )
+       RETURN
     ENDIF
-#endif
+
+#else
```

Comment:
The rest of this seems to be the original code, and the above is a duplicate of it with some modifications. We need to better put these two implementations together to avoid so much duplicate code.